File: ios/SuperheroApp/CharacterDetail/UseCase/GetCharacterUseCase.swift
//
//  GetCharacterUseCase.swift
//  SuperheroApp
//
//  Created by Balazs Varga on 2018. 02. 12..
//  Copyright © 2018. W.UP. All rights reserved.
//

import Common

class GetCharacterRequest: CommonUseCaseRequest {
    let characterId: Int32

    init(characterId: Int32) {
        self.characterId = characterId
    }
}

class GetCharacterResponse: CommonUseCaseResponse {
    let character: CommonCharacter?

    init(character: CommonCharacter?) {
        self.character = character
    }
}

class GetCharacterUseCase: CommonUseCase {

    private var charactersDataSource: CommonCharactersDataSource

    init(charactersDataSource: CommonCharactersDataSource) {
        self.charactersDataSource = charactersDataSource
    }

    override func executeUseCase(request: CommonUseCaseRequest) {

        if let request = request as? GetCharacterRequest {

            self.charactersDataSource.loadCharacter(characterId: request.characterId, complete: { (character) in
                let response = GetCharacterResponse(character: character)
                return self.success(response)
            }, fail: {
                return self.error()
            })
        }
    }
}

File: common/src/main/kotlin/digital/wup/superheroapp/common/charaterdetail/domain/usecase/GetCharacterUseCase.kt
package digital.wup.superheroapp.common.charaterdetail.domain.usecase

import digital.wup.superheroapp.common.UseCase
import digital.wup.superheroapp.common.UseCaseRequest
import digital.wup.superheroapp.common.UseCaseResponse
import digital.wup.superheroapp.common.characters.domain.model.Character
import digital.wup.superheroapp.common.datasource.CharactersDataSource

class GetCharacterRequest(val characterId: Int) : UseCaseRequest

class GetCharacterResponse(val character: Character?) : UseCaseResponse

class GetCharacterUseCase(private val charactersDataSource: CharactersDataSource) : UseCase<GetCharacterRequest, GetCharacterResponse>() {

    override fun executeUseCase(request: GetCharacterRequest) {
        charactersDataSource.loadCharacter(request.characterId, {
            val response = GetCharacterResponse(it)
            success(response)
        }, {
            error()
        })
    }
}